### PR TITLE
fix(client): Fix client using localhost for chunkservers

### PR DESCRIPTION
**Pull request by uristdwarf**
When master gives information regarding chunks and chunkservers, it may give the localhost ip addresses of chunkserver to client, causing the client to search for the chunks on their localhost. This commit fixes that by interpreting localhost from master to mean the master IP address.

### CODE REVIEW

File: src/mount/mastercomm.cc [hunk 1]
```diff
@@ -2118,6 +2121,13 @@ uint8_t fs_saureadchunk(std::vector<ChunkTypeWithAddress> &chunkservers, uint64_
 		setDisconnect(true);
 		return SAUNAFS_ERROR_IO;
 	}
+	for (auto& server : chunkservers) {
+		// If 127.0.0.1, let's assume it's the same as master
+		if (server.address.ip == localhost) {
+			server.address.ip=masterip;
+			safs::log_debug("changing chunkserver ip address 127.0.0.1 to {}", htonl(masterip));
+		}
+	}
```
Comment (rolysr) [inline src/mount/mastercomm.cc:2130]: As a minor suggestion consider for readability to encapsulate this logic into a function for reusing it.

File: src/mount/mastercomm.cc [hunk 2]
```diff
@@ -2213,6 +2223,13 @@ uint8_t fs_sauwritechunk(uint32_t inode, uint32_t chunkIndex, uint32_t &lockId,
 		setDisconnect(true);
 		return SAUNAFS_ERROR_IO;
 	}
+	for (auto& server : chunkservers) {
+		// If 127.0.0.1, let's assume it's the same as master
+		if (server.address.ip == localhost) {
+			server.address.ip=masterip;
```
Comment (rolysr) [inline src/mount/mastercomm.cc:2229]: Consider adding a space in this line to both sides of the `=` .

**Review by rolysr — COMMENTED** [review]
LGTM. But consider my minor suggestions for code readabilty. On the other hand, did you considered if it is necessary to add a test for checking the fix of the issue that was happening before? Just for checking there is not edge case failing for this.

File: src/mount/mastercomm.cc (hunk 2 above)
Comment (antuan96314) [inline src/mount/mastercomm.cc:2229]:
```suggestion
			server.address.ip = masterip;
```

**Review by antuan96314 — DISMISSED** [review]

**Review by aNeutrino — CHANGES_REQUESTED** [review]

File: src/mount/mastercomm.cc (hunk 1 above)
[the remark this answers is not in the document]
Comment (uristdwarf) [reply src/mount/mastercomm.cc]: > and some test ?

I wasn't sure how to actually do this test, because our testing framework doesn't use local ip addresses (master never supported them until recently), but I'll try the test you send out. If it fails, I'll just write a unit test with some master mocking (but it might take a while).

> why MASTER is returning locahost on the FIRSTPLACE ?

Because master stores the chunkserver IP address used to connect and gives this directly to client, and that's it. Previously chunkserver failed to connect at all if it was on localhost, but that has changed.

[the remark this answers is not in the document]
Comment (uristdwarf) [reply src/mount/mastercomm.cc]: > to be honest I do not know what is a need/goal of your code change ....

Simple, allow client to connect to chunk servers that use the same localhost IP address as master.

Comment (uristdwarf) [reply src/mount/mastercomm.cc]: Done

File: src/mount/mastercomm.cc (hunk 2 above)
Comment (uristdwarf) [reply src/mount/mastercomm.cc]: Done

**Review by ralcolea — APPROVED** [review]
LGTM. Please, consider adding an automated test where at least one chunkserver use localhost as IP to validate the change.

**Comment by uristdwarf** [comment]
> LGTM. Please, consider adding an automated test where at least one chunkserver use localhost as IP to validate the change.

I tried adding one, but it isn't simple (with our current testing framework). This problem occurs because of a client on a different host.

You have chunkserver/master on host A, and client on host B. Master returns the chunkserver ip address relative to it's own address on host B. However with the testing framework everything is tested on the single host, where this bug doesn't happen. If I researched this more I'm sure I could find a way to emulate this behavior, but I don't think it's worth the time or effort

**Comment by uristdwarf** [comment]
> If I researched this more I'm sure I could find a way to emulate this behavior, but I don't think it's worth the time or effort

At least not for this PR, which is my personal contribution I made because I found and fixed it in my spare time.

**Review by antuan96314 — APPROVED** [review]
